src/server/context.ts

```diff
@@ -78,75 +82,153 @@ function requestMotokoInstance(
-                }
-                return require('motoko/lib').default(compiler);
+): Promise<[Motoko, MocJsInfo]> {
+    const configuredMocJsPath = settings.mocJsPath;
+    if (configuredMocJsPath) {
+        console.log('Trying to use custom moc.js path:', configuredMocJsPath);
+        const resolvedPath = isAbsolute(configuredMocJsPath)
+            ? configuredMocJsPath
+            : resolve(workspaceRoot, configuredMocJsPath);
+        const filename = basename(resolvedPath);
+        const mocVersion = extractMocVertion(filename);
+
+        if (existsSync(resolvedPath)) {
+            const res = getCompiler(resolvedPath);
+            if (res.isOk()) {
+                const [compiler, path] = res.value;
+                return [
+                    require('motoko/lib').default(compiler),
+                    {
+                        // assume that moc.js version in file name is correct, otherwise
+                        // use provided version
+                        version: mocVersion || version,
+                        path,
+                    },
+                ];
             } else {
                 // TODO: notify user via `connection.sendMessage()`
                 console.error(
-                    `Custom moc.js file not found at ${resolvedPath}, falling back to default`,
+                    `Error loading custom moc.js from ${configuredMocJsPath}:`,
+                    res.error.message,
                 );
             }
-        } catch (error: any) {
+        } else {
             // TODO: notify user via `connection.sendMessage()`
-            console.error(
-                `Error loading custom moc.js from ${mocJsPath}:`,
-                error,
+            console.warn(
+                `Custom moc.js file not found at ${resolvedPath}, trying to download`,
             );
+            if (mocVersion) {
+                const res = await getMocJs(
+                    mocVersion,
+                    dirname(resolvedPath),
+                ).andThen((path) => getCompiler(path));
+                if (res.isOk()) {
+                    const [compiler, path] = res.value;
+                    return [
+                        require('motoko/lib').default(compiler),
+                        {
+                            version: mocVersion,
+                            path,
+                        },
+                    ];
+                } else {
+                    console.error(
+                        `Error while downloading custom moc.js to ${configuredMocJsPath}:`,
+                        res.error.message,
+                    );
+                }
+            }
         }
     }
 
-    if (version === '0.10.4') {
-        const compiler = require(join(
-            __dirname,
-            '/compiler/moc-' + version,
-        )).Motoko;
-        return require('motoko/lib').default(compiler);
+    if (version) {
+        const mocJsDir = join(__dirname, 'compiler');
```

Comment:
So, if no path is configured, download it to `<current working directory>/compiler`? Perhaps print a message informing the user about it.

One thing, we've been mostly using `console.log`. Do you think it would make sense to use `vscode.window.showInformationMessage` here, perhaps?

Reply:
The replaced code and analysis of package.json shows that when the extension runs the `__dirname` == `./out`. That means that compiler should be downloaded into internal extension's directory. I don't think that the user should know much about those internals.